core/modules/wpublish/QueryChunkStatistics.cc

```diff
@@ -210,6 +211,53 @@ std::ostream& operator<<(std::ostream& os, QueryStatistics const& q) {
 }
 
 
+/// Remove all Tasks belonging to the user query 'qId' from the queue of 'sched'.
+/// If 'sched' is null, then Tasks will be removed from the queue of whatever scheduler
+/// they are queued on.
+/// Tasks that are already running continue, but are marked as complete on their
+/// current scheduler. Stopping and rescheduling them would be difficult at best, and
+/// probably not very helpful.
+std::vector<wbase::Task::Ptr>
+QueryChunkStatistics::removeQueryFrom(QueryId const& qId,
+                                      std::shared_ptr<wsched::SchedulerBase> const& sched) {
+
+    std::vector<wbase::Task::Ptr> removedList; // Return value;
+
+    // Find the user query.
+    std::unique_lock<std::mutex> lock(_queryStatsMtx);
+    auto query = _queryStats.find(qId);
+    if (query == _queryStats.end()) {
+        LOGS(_log, LOG_LVL_DEBUG, QueryIdHelper::makeIdStr(qId) << " was not found by removeQueryFrom");
+        return removedList;;
+    }
+    lock.unlock();
+
+    // Remove Tasks from their scheduler put them on 'removedList', but only if their Scheduler is the same
+    // as 'sched'or if sched == nullptr.
+    auto &taskMap = query->second->_taskMap;
+    std::vector<wbase::Task::Ptr> taskList;
+    {
+        std::lock_guard<std::mutex> taskLock(query->second->_qStatsMtx);
+        for (auto const& elem : taskMap) {
+            taskList.push_back(elem.second);
+        }
+    }
+
+    for(auto const& task : taskList) {
+        auto baseTaskSched = std::dynamic_pointer_cast<wsched::SchedulerBase>(task->getTaskScheduler());
```

Comment:
It could return nullptr. There is a brief delay between a task being received and it being placed on a scheduler. Adding TaskScheduler::removeTask is reasonable.